src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt

```diff
@@ -82,6 +95,70 @@ class RsFileStub : PsiFileStubImpl<RsFile> {
     }
 }
 
+/**
+ * A utility used in the stub builder to detect that a code block [RsBlockStubType] should not be
+ * traversed in order to find child stubs. Avoiding code blocks traversing is important because they
+ * are lazy parseable and will not be parsed during indexing if not traversed.
+ * The utility is heuristic based, i.e. it can strictly say that a block does NOT have stubs, but can't
+ * strictly say that block has stubs.
+ *
+ * There are two functions that do mostly the same but differs in how they use the cache.
+ * - [computeAndCache] computes the value and then puts it in the cache. Used in [RsFunctionStub.Type.createStub]
+ * - [getAndClearCached] retrieves and removes the value from the cache or compute the value if it does not
+ *   present. Expected to be called after [computeAndCache] during indexing (because a function is a parent of a
+ *   block, so it is processed earlier)
+ *
+ * See tests in `RsLazyCodeBlockIsNotExpandedDuringStubBuildingTest`
+ */
+private object BlockCanHaveStubsHeuristic {
```

Comment:
Maybe `BlockStubUtils` name is better here?
Just to avoid some complex name and pointing to implementation details